packages/vscode-extension/src/project/deviceSession.ts

```diff
@@ -111,10 +117,30 @@ export class DeviceSession implements Disposable {
 
     Logger.debug("Will wait for app ready and for preview");
     this.eventDelegate.onStateChange(StartupMessage.WaitingForAppToLoad);
+
+    if (shouldWaitForAppLaunch) {
+      const reportWaitingStuck = setTimeout(() => {
+        Logger.info("App is taking very long to boot up, it might be stuck");
+        getTelemetryReporter().sendTelemetryEvent("app:launch:waiting-stuck", {
+          platform: this.device.platform,
+        });
+      }, 30000);
+      waitForAppReady.then(() => clearTimeout(reportWaitingStuck));
+    }
+
     const [previewUrl] = await Promise.all([this.device.startPreview(), waitForAppReady]);
```

Comment:
Shouldn't we wait for waitForAppReady once?
```suggestion
    }

    const [previewUrl] = await Promise.all([this.device.startPreview(), waitForAppReady]);
    if (shouldWaitForAppLaunch) {
       clearTimeout(reportWaitingStuck);
    }
```

Reply:
I like my approach moare as it doesn't mix the reporting code with the existing logic. Whatever is placed inside that if block I added, is just for reporting and can be deleted. Otherwise I'd have to keep the `reportWaitingStuck` in main scope, while we'd need to have `shouldWaitForAppLaunch` checks both before and after the await.

On top of that this isn't really an equivalent, as the await is for both startPreview and app ready event.